### DISCUSSION TITLE
How do I unzip the zip file?

### DISCUSSION THREAD

**Discussion by vladstudio** [mikekazakov/nimble-commander#141] [Q&A] [answered]
I might be missing something obvious. Here's what I tried:

- opening the zip file with Enter. Result: zip file content opens as a folder.
- right-clicking to open context menu. No options there to uncompress.
- searched Hotkeys for uncompress, extract.

As a workaround, I added a "tool" for The Unarchiver app. Is this the only way?

**Comment by mikekazakov** [comment]
Right now the only way to extract items from an archive is to open it as a directory, select everything and then copy into destination.

**Comment by WWWFWWW** [comment, marked as answer]
> opening the zip file with Enter. Result: zip file content opens as a folder

You can always use _Open_ instead of _Enter_. IIRC the shortcut is simply ⇧⏎ instead of only ⏎

This is similar behaviour as when opening or entering application or other bundles from inside NC (Something.app "files", which are actually folders). _Enter_ only browses the .app folder inside the file panel (as it should, IMHO), _Open_ launches the app.

⇧⏎ will open the file in its Finder‑assigned application, which by default is Archive Utility app set to extract the whole archive to a new folder, just like double‑clicking the zip in Finder.

Or the same as right‑clicking the zip file in NC and choosing Open (or Open With... for a list of compatible applications, if you have more unarchivers selected).

IMHO entering archives as folders in orthodox file commanders is the expected behaviour, as people either want to view or extract just individual files, or select all inside it and copy it to any folder in the other file panel, as orthodox file managers are all about having two file panels open for any file operations.

Consider how easier is it to just browse an archive in the left panel, selecting individual or all files inside and copying them to directly to your set destination in the right panel, instead of extracting the archive in place, navigating to its extracted location and copying/moving the extracted files to your set destination. And for the cases where extract in place is needed, easily opening the archive with Archive Utility works well enough for me. Though sure, it can be somewhat counterintuitive when one expects an "Extract" command ;-)

**Reply by mikekazakov** [reply]
Thank you for the well-written summary!
I'd only add that while everything here is true, there's still a valid reason for a separate "unarchive" action to exist - and that is performance. It's much easier to tailor the code specifically for that, instead of using a generic operation which copies generic items from a generic virtual file system...

**Reply by WWWFWWW** [reply]
Ah, yes – it makes sense that there would be some overhead if working with archives as VFS.

Out of interest (sorry, can't delve into NC's source that much at the moment) – what compression/decompression library do you use in NC? I remember that you added a bunch of newly supported archive formats a while back, which was nice. 

I'd still say that for the simple decompression of a whole archive, especially of a big one, I still prefer opening it with a separate application like Archive Utility – as far as I could tell, the whole Apple Archive suite of tools got some pretty nice optimisations for multithreading, Apple Silicon and native support for stuff like APFS sparse bundles and clone files, at least with Apple's archive format, or even setting different QoS for the compression and decompression (i.e. I might want to decompress using Performance cores, for fastest decompression, but compress using Efficient cores, to be able to work on something else while a huge dataset compresses efficiently in the background).

Perhaps all of that could be – or maybe some already is – implemented in NC, but I am quite content to use other apps for that, especially if they can be made to work with NC pretty seamlessly (and I can still use the Tools feature to set up native commandline tools like "aa" for more control).

IMHO having a separate _unarchive_ function could still be nice if faster than accessing archives via VFS to decompress the whole archive, but if it means switching libraries or refactoring the code significantly, one has to ask if it's worth it – when pretty much the same thing could be possibly done by just adding a Tools entry to NC with something like

taskpolicy -c background -s /usr/bin/aa &arg:extract &arg:-i &arg:%p &arg:-d &arg:%-r &arg:-keychain &arg:-afsc-all &arg:-t &arg:4

_(WARNING: that's really a very mangled pseudo‑command, so **don't use as it is**, as it absolutely won't work and might rm rf / bork your whole computer – look up the aa manual and format a proper command yourself! Taskpolicy is not that well documented and I haven't really checked how it parses the different &arg: needed, if at all!)_

to hopefully extract the selected archive to the working directory in the inactive file panel, using E cores in "background" QoS policy, loading password from the keychain and keeping the extracted files compressed for on the fly decompression on APFS disks (as dumb as that last bit sounds). Of course, doing it without taskpolicy would be even more straightforward.

And simply opening the archive with Archive Utility to decompress it into the working directory (as is its default) usually works for me in most cases as well.

**Reply by mikekazakov** [reply]
NC uses libarchive under the hood: https://github.com/libarchive/libarchive
